File: LLD/src/main/java/org/example/p1_oops/c7_composition/association/Student.java
package org.example.p1_oops.c7_composition.association;

// Association: Student HAS-A Teacher, but both exist INDEPENDENTLY
public class Student {
    private String name;
    private Teacher teacher; // Reference to Teacher (passed from outside)

    public Student(String name, Teacher teacher) {
        this.name = name;
        this.teacher = teacher; // Teacher is NOT created here, just referenced
    }

    public void study() {
        System.out.println("Student " + name + " is studying " + teacher.getSubject() + " with " + teacher.getName());
    }

    // Teacher can exist without Student
    // Student can change Teacher
    // This is ASSOCIATION (weak relationship)
}

File: LLD NodeJs/p1_oops/c7_composition/association/Student.js
/**
 * Student class demonstrating Association (weak HAS-A relationship)
 * 
 * Association: Student HAS-A Teacher, but both exist INDEPENDENTLY.
 * The Teacher is passed from outside, not created inside.
 */
class Student {
    #name;
    #teacher; // Reference to Teacher (passed from outside)

    constructor(name, teacher) {
        this.#name = name;
        this.#teacher = teacher; // Teacher is NOT created here, just referenced
    }

    study() {
        console.log("Student " + this.#name + " is studying " +
            this.#teacher.getSubject() + " with " + this.#teacher.getName());
    }

    // Teacher can exist without Student
    // Student can change Teacher
    // This is ASSOCIATION (weak relationship)
}

module.exports = Student;
